src/drivers/irlock/irlock.cpp

```diff
@@ -0,0 +1,410 @@
+#endif
+
+class IRLOCK : public device::I2C
+{
+public:
+	IRLOCK(int bus = IRLOCK_I2C_BUS, int address = IRLOCK_I2C_ADDRESS);
+	virtual ~IRLOCK();
+
+	virtual int init();
+	virtual int probe();
+	virtual int info();
+
+	virtual ssize_t read(struct file *filp, char *buffer, size_t buflen);
+
+private:
+
+	/* start periodic reads from sensor */
+	void start();
+
+	/* stop periodic reads from sensor */
+	void stop();
+
+	/* static function that is called by worker queue, arg will be pointer to instance of this class */
+	static void		cycle_trampoline(void *arg);
+
+	/* read from device and schedule next read */
+	void			cycle(); \
+
+	/* low level communication with sensor */
+	int read_device();
+	bool sync_device();
+	int read_device_word(uint16_t *word);
+	int read_device_block(struct irlock_s *block);
+
+	/* internal variables */
+	RingBuffer *_reports;
+	bool _sensor_ok;
+	work_s _work;
+};
+
+/* global pointer for single IRLOCK sensor */
+namespace
+{
+IRLOCK *g_irlock = nullptr;
+}
+
+void irlock_usage();
+
+extern "C" __EXPORT int irlock_main(int argc, char *argv[]);
+
+/* constructor */
+IRLOCK::IRLOCK(int bus, int address) :
+	I2C("irlock", IRLOCK0_DEVICE_PATH, bus, address, 400000),
+	_reports(nullptr),
+	_sensor_ok(false)
+{
+	memset(&_work, 0, sizeof(_work));
+}
+
+/* destructor */
+IRLOCK::~IRLOCK()
+{
+	stop();
+
+	/* clear reports queue */
+	if (_reports != nullptr) {
+		delete _reports;
+	}
+}
+
+/* initialise driver to communicate with sensor */
+int IRLOCK::init()
+{
+	/* initialise I2C bus */
+	int ret = I2C::init();
+
+	if (ret != OK) {
+		return ret;
+	}
+
+	/* allocate buffer storing values read from sensor */
+	_reports = new RingBuffer(135, sizeof(struct irlock_s));
+
+	if (_reports == nullptr) {
+		return ENOTTY;
+
+	} else {
+		_sensor_ok = true;
+		/* start work queue */
+		start();
+		return OK;
+	}
+}
+
+/* probe the device is on the I2C bus */
+int IRLOCK::probe()
+{
+	/*
+	 * IRLock defaults to sending 0x00 when there is no block
+	 * data to return, so really all we can do is check to make
+	 * sure a transfer completes successfully.
+	 */
+	uint8_t byte;
+
+	if (transfer(nullptr, 0, &byte, 1) != OK) {
+		return -EIO;
+	}
+
+	return OK;
+}
+
+/* display driver info */
+int IRLOCK::info()
+{
+	if (g_irlock == nullptr) {
+		errx(1, "irlock device driver is not running");
+	}
+
+	/* display reports in queue */
+	if (_sensor_ok) {
+		_reports->print_info("report queue: ");
+	} else {
+		warnx("sensor is not healthy");
+	}
+
+	return OK;
+}
+
+/* start periodic reads from sensor */
+void IRLOCK::start()
+{
+	// flush ring and reset state machine
+	_reports->flush();
+
+	// start work queue cycle
+	work_queue(HPWORK, &_work, (worker_t)&IRLOCK::cycle_trampoline, this, 1);
+}
+
+/* stop periodic reads from sensor */
+void IRLOCK::stop()
+{
+	work_cancel(HPWORK, &_work);
+}
+
+void IRLOCK::cycle_trampoline(void *arg)
+{
+	IRLOCK *device = (IRLOCK *)arg;
+
+	/* check global irlock reference and cycle */
```

Comment:
The other comments in the file use C++ style (```//```), so it would be good to change the in-function comments to one consistent style.